keystore/keystore.go

```diff
@@ -0,0 +1,222 @@
+
+	gethkeystore "github.com/ethereum/go-ethereum/accounts/keystore"
+	gethcrypto "github.com/ethereum/go-ethereum/crypto"
+	"github.com/smartcontractkit/chainlink-common/keystore/internal"
+	"github.com/smartcontractkit/chainlink-common/keystore/serialization"
+	"github.com/smartcontractkit/chainlink-common/keystore/storage"
+	"google.golang.org/protobuf/proto"
+)
+
+type KeyType string
+
+const (
+	X25519 KeyType = "X25519"
+	// TODO: Support P256 for DKG.
+	// Digital signature key types.
+	Ed25519        KeyType = "ed25519"
+	EcdsaSecp256k1 KeyType = "ecdsa-secp256k1"
+)
+
+var AllKeyTypes = []KeyType{X25519, Ed25519, EcdsaSecp256k1}
+
+type ScryptParams struct {
+	N int
+	P int
+}
+
+var (
+	DefaultScryptParams = ScryptParams{
+		N: gethkeystore.StandardScryptN,
+		P: gethkeystore.StandardScryptP,
+	}
+	FastScryptParams ScryptParams = ScryptParams{
+		N: 1 << 14,
+		P: 1,
+	}
+)
+
+// KeyInfo is the information about a key in the keystore.
+// Public key may be empty for non-asymmetric key types.
+type KeyInfo struct {
+	Name      string
+	KeyType   KeyType
+	CreatedAt time.Time
+	PublicKey []byte
+	Metadata  []byte
+}
+
+// newKeyInfo is a private constructor that ensures all fields are set explicitly.
+func newKeyInfo(name string, keyType KeyType, createdAt time.Time, publicKey []byte, metadata []byte) KeyInfo {
+	return KeyInfo{
+		Name:      name,
+		KeyType:   keyType,
+		CreatedAt: createdAt,
+		PublicKey: publicKey,
+		Metadata:  metadata,
+	}
+}
+
+type Keystore interface {
+	Admin
+	Reader
+	Signer
+	Encryptor
+}
+
+type key struct {
+	keyType    KeyType
+	privateKey internal.Raw
+	createdAt  time.Time
+	metadata   []byte
+
+	// Derived from private key on loading from storage.
+	// Cached here for convenience.
+	publicKey []byte
+}
+
+// newKey is a private constructor ensuring the internal key struct is fully and correctly populated.
+func newKey(keyType KeyType, privateKey internal.Raw, publicKey []byte, createdAt time.Time, metadata []byte) key {
+	return key{
+		keyType:    keyType,
+		privateKey: privateKey,
+		publicKey:  publicKey,
+		createdAt:  createdAt,
+		metadata:   metadata,
+	}
+}
+
+// EncryptionParams controls password-based encryption cost.
+// N and P are scrypt parameters; higher values increase CPU/memory cost.
+// Password is the secret used to derive the encryption key.
+type EncryptionParams struct {
+	Password     string
+	ScryptParams ScryptParams
+}
+
+func publicKeyFromPrivateKey(privateKeyBytes internal.Raw, keyType KeyType) ([]byte, error) {
+	switch keyType {
+	case Ed25519:
+		return ed25519.PublicKey(internal.Bytes(privateKeyBytes)), nil
+	case EcdsaSecp256k1:
+		// Here we use SEC1 (uncompressed) format for ECDSA public keys.
+		// Its commonly used and EVM addresses are derived from this format.
+		// We use the geth crypto library for secp256k1 support
+		// because stdlib doesn't support it.
+		privateKey, err := gethcrypto.ToECDSA(internal.Bytes(privateKeyBytes))
```

Comment:
I'm not sure why the core node logic doesn't use these geth methods but instead copies the inner logic (still depends on geth so wouldn't be to remove the dep) https://github.com/smartcontractkit/chainlink/blob/develop/core/services/keystore/keys/ethkey/key_v2.go#L23. Seems cleaner to use the methods unless I'm missing something